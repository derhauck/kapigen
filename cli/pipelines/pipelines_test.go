package pipelines

import (
	"bytes"
	"os"
	"strings"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"gitlab.com/kateops/kapigen/cli/types"
	"gitlab.com/kateops/kapigen/dsl/gitlab/job"
	"gitlab.com/kateops/kapigen/dsl/gitlab/pipeline"
	"gitlab.com/kateops/kapigen/dsl/gitlab/stages"
	"gopkg.in/yaml.v3"
)

func TestCreatePipeline(t *testing.T) {
	t.Run("can create pipeline", func(t *testing.T) {
		file := "pipeline.yaml"
		CreatePipeline(func(jobs *types.Jobs, ciPipeline *pipeline.CiPipeline) {
			ciPipeline.DefaultCiPipeline()
			jobs.AddJob(types.NewJob("generic", "alpine", func(ciJob *job.CiJob) {
				ciJob.TagMediumPressure().
					AddScript("echo hello world").
					SetStage(stages.TEST).
					Rules.AddRules(*job.DefaultMainBranchRules())
				ciJob.Secrets = job.Secrets{
					"TEST": &job.VaultSecret{
						Vault: job.VaultSecretConfig{
							Engine: job.VaultSecretEngine{
								Name: job.EnumVaultSecretEngineKv2,
								Path: "mount",
							},
							Path:  "path",
							Field: "field",
						},
						Token: "token",
					},
				}
			}))
		})
		readFile, err := os.ReadFile(file)
		if err != nil {
			t.Error(err)
		}
		pipelineConfig := map[string]any{}
		err = yaml.NewDecoder(bytes.NewReader(readFile)).Decode(&pipelineConfig)
		if err != nil {
			t.Error(err)
		}
		snaps.MatchSnapshot(t, pipelineConfig["generic"], pipelineConfig["variables"], pipelineConfig["workflow"], pipelineConfig["default"])
		snaps.MatchSnapshot(t, string(readFile))
		err = os.Remove(file)
		if err != nil {
			t.Error(err)
		}

	})
	t.Run("can not create pipeline", func(t *testing.T) {
		file := "pipeline.yaml"
		_ = os.Remove(file)
		CreatePipeline(func(jobs *types.Jobs, ciPipeline *pipeline.CiPipeline) {
			jobs.AddJob(types.NewJob("invalid", "alpine", func(ciJob *job.CiJob) {
			}))
		})
		_, err := os.ReadFile(file)
		if err == nil {
			t.Errorf("should not be able to open %s", file)
			_ = os.Remove(file)
		}

		if err.Error() != "open pipeline.yaml: no such file or directory" {
			t.Errorf("expectec: %s, received: %s", err, "open pipeline.yaml: no such file or directory")
		}

	})
}

func TestReadPipelineConfig(t *testing.T) {
	t.Run("can create pipelineConfig", func(t *testing.T) {
		configPath := "../config.kapigen.yaml"
		config, err := ReadPipelineConfig(configPath)
		if err != nil {
			t.Error(err)
		}
		if config.Noop == false {
			t.Error("should be true")
		}

		if config.Versioning == false {
			t.Error("should be true")
		}

		if config.PrivateTokenName != "CI_PIPELINE_TOKEN" {
			t.Error("should be CI_PIPELINE_TOKEN")
		}

	})
	t.Run("can not create pipelineConfig", func(t *testing.T) {
		configPath := "config.kapigen.yaml-missing"
		config, err := ReadPipelineConfig(configPath)
		expectedError := "open config.kapigen.yaml-missing: no such file or directory"
		if config != nil {
			t.Error("config should be nil")
		}
		if err == nil {
			t.Error("should contain error")
			t.FailNow()
		}
		if !strings.Contains(err.Error(), expectedError) {
			t.Errorf("expected: '%s', received: '%s'", expectedError, err)
		}

	})
}
