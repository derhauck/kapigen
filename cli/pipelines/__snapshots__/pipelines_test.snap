
[TestCreatePipeline/can_create_pipeline - 1]
map[string]interface {}{
    "allow_failure": bool(false),
    "image":         map[string]interface {}{
        "name":        "alpine",
        "pull_policy": "always",
    },
    "needs": []interface {}{
    },
    "rules": []interface {}{
        map[string]interface {}{
            "allow_failure": bool(false),
            "if":            "$KTC_STOP_PIPELINE != \"false\" && $DEBUG == null",
            "when":          "never",
        },
        map[string]interface {}{
            "allow_failure": bool(false),
            "if":            "$CI_DEFAULT_BRANCH == $CI_COMMIT_BRANCH",
            "when":          "on_success",
        },
        map[string]interface {}{
            "allow_failure": bool(false),
            "if":            "$KTC_TEST_PIPELINE",
            "when":          "on_success",
        },
    },
    "script": []interface {}{
        "echo hello world",
    },
    "secrets": map[string]interface {}{
        "TEST": map[string]interface {}{
            "token": "token",
            "vault": map[string]interface {}{
                "engine": map[string]interface {}{
                    "name": "kv-v2",
                    "path": "mount",
                },
                "field": "field",
                "path":  "path",
            },
        },
    },
    "services": []interface {}{
    },
    "stage": "dynamic",
    "tags":  []interface {}{
        "pressure:medium",
    },
}
map[string]interface {}{
    "KTC_STOP_PIPELINE": "false",
}
map[string]interface {}{
    "name":  "default",
    "rules": []interface {}{
        map[string]interface {}{
            "if":   "$CI_MERGE_REQUEST_ID",
            "when": "always",
        },
        map[string]interface {}{
            "if":   "$CI_DEFAULT_BRANCH == $CI_COMMIT_BRANCH",
            "when": "always",
        },
        map[string]interface {}{
            "if":   "$CI_MERGE_REQUEST_IID && $CI_PIPELINE_SOURCE == 'merge_request_event'",
            "when": "always",
        },
        map[string]interface {}{
            "if":   "$CI_COMMIT_TAG != null",
            "when": "always",
        },
    },
}
nil
---

[TestCreatePipeline/can_create_pipeline - 2]
default: null
generic:
    allow_failure: false
    script:
        - echo hello world
    needs: []
    image:
        name: alpine
        pull_policy: always
    rules:
        - if: $KTC_STOP_PIPELINE != "false" && $DEBUG == null
          allow_failure: false
          when: never
        - if: $CI_DEFAULT_BRANCH == $CI_COMMIT_BRANCH
          allow_failure: false
          when: on_success
        - if: $KTC_TEST_PIPELINE
          allow_failure: false
          when: on_success
    stage: dynamic
    services: []
    tags:
        - pressure:medium
    secrets:
        TEST:
            vault:
                engine:
                    name: kv-v2
                    path: mount
                path: path
                field: field
            token: token
stages:
    - lint
    - init
    - build
    - test
    - release
    - dynamic
    - trigger
    - final
variables:
    KTC_STOP_PIPELINE: "false"
workflow:
    name: default
    rules:
        - if: $CI_MERGE_REQUEST_ID
          when: always
        - if: $CI_DEFAULT_BRANCH == $CI_COMMIT_BRANCH
          when: always
        - if: $CI_MERGE_REQUEST_IID && $CI_PIPELINE_SOURCE == 'merge_request_event'
          when: always
        - if: $CI_COMMIT_TAG != null
          when: always

---
